File: backend/internal/models/post.go
package models

import "time"

type Post struct {
	ID              int        `json:"id"`
	TopicID         int        `json:"topic_id"`
	Title           string     `json:"title"`
	Summary         string     `json:"summary"`
	Content         string     `json:"content"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`
	UserID          int        `json:"user_id"`
	PinnedCommentID *int       `json:"pinned_comment_id,omitempty"`
	Score           int        `json:"score"`
	NoOfComments    int        `json:"no_of_comments"`
	IsDeleted       bool       `json:"is_deleted"`
	DeletedAt       *time.Time `json:"deleted_at,omitempty"`
	MyVote          int        `json:"my_vote,omitempty"`
	TopicName       string     `json:"topic_name,omitempty"`
	Username        string     `json:"username,omitempty"`
}

type PostVote struct {
	UserID    int `json:"user_id"`
	PostID    int `json:"post_id"`
	VoteValue int `json:"vote_value"`
}

type CreatePostRequest struct {
	TopicID int    `json:"topic_id" schema:"topic_id"`
	Title   string `json:"title" schema:"title"`
	Content string `json:"content" schema:"content"`
}

type UpdatePostRequest struct {
	Title   string `json:"title,omitempty" schema:"title"`
	Content string `json:"content,omitempty" schema:"content"`
}

type VotePostRequest struct {
	VoteValue int `json:"vote_value" schema:"vote_value"`
}

type ListPostsRequest struct {
	Page                 int    `json:"page,omitempty" schema:"page"`
	PageSize             int    `json:"page_size,omitempty" schema:"page_size"`
	Sort                 string `json:"sort,omitempty" schema:"sort"`
	OrderBy              string `json:"order_by,omitempty" schema:"order_by"`
	Search               string `json:"search,omitempty" schema:"search"`
	TopicID              *int   `json:"topic_id,omitempty" schema:"topic_id"`
	UserID               *int   `json:"user_id,omitempty" schema:"user_id"`
	FilterFollowingTopics bool   `json:"filter_following_topics,omitempty" schema:"filter_following_topics"`
	ShowDeletedPosts     bool   `json:"show_deleted_posts,omitempty" schema:"show_deleted_posts"`
}

type PinCommentRequest struct {
	CommentID *int `json:"comment_id" schema:"comment_id"`
}

File: frontend/src/types/post.ts
import type { z } from "zod";
import type {
  createPostSchema,
  listPostsRequestSchema,
  listPostsSearchParamsSchema,
  postsSortValueSchema,
  updatePostSchema,
} from "@/schema/posts";

export interface Post {
  id: number;
  topic_id: number;
  title: string;
  summary: string;
  content: string;
  created_at: string;
  updated_at: string;
  user_id: number;
  pinned_comment_id: number | null;
  score: number;
  no_of_comments: number;
  is_deleted: boolean;
  deleted_at: string | null;
  my_vote: number;
  topic_name: string;
  username: string;
}

export type PostsSortValue = z.infer<typeof postsSortValueSchema>;

export type ListPostsSearchParams = z.infer<typeof listPostsSearchParamsSchema>;

export type ListPostsRequest = z.infer<typeof listPostsRequestSchema>;
export type CreatePostRequest = z.infer<typeof createPostSchema>;
export type UpdatePostRequest = z.infer<typeof updatePostSchema>;

export type VotePostRequest = {
  vote_value: -1 | 0 | 1;
};

export type PinCommentRequest = {
  comment_id: number | null;
};

export interface PaginatedPostsResponse {
  posts: Post[];
  count: number;
}
